src/common/os_input_output.rs

```diff
@@ -136,64 +151,80 @@ fn spawn_terminal(file_to_open: Option<PathBuf>, orig_termios: Termios) -> (RawF
 
 #[derive(Clone)]
 pub struct OsInputOutput {
-    orig_termios: Arc<Mutex<Termios>>,
+    orig_termios: Arc<Mutex<termios::Termios>>,
 }
 
+/// The `OsApi` trait represents an abstract interface to the features of an operating system that
+/// Zellij requires.
 pub trait OsApi: Send + Sync {
-    fn get_terminal_size_using_fd(&self, pid: RawFd) -> PositionAndSize;
-    fn set_terminal_size_using_fd(&mut self, pid: RawFd, cols: u16, rows: u16);
-    fn set_raw_mode(&mut self, pid: RawFd);
-    fn unset_raw_mode(&mut self, pid: RawFd);
+    /// Returns the size of the terminal associated to file descriptor `fd`.
+    fn get_terminal_size_using_fd(&self, fd: RawFd) -> PositionAndSize;
+    /// Sets the size of the terminal associated to file descriptor `fd`.
+    fn set_terminal_size_using_fd(&mut self, fd: RawFd, cols: u16, rows: u16);
+    /// Set the terminal associated to file descriptor `fd` to
+    /// [raw mode](https://en.wikipedia.org/wiki/Terminal_mode).
+    fn set_raw_mode(&mut self, fd: RawFd);
+    /// Set the terminal associated to file descriptor `fd` to
+    /// [cooked mode](https://en.wikipedia.org/wiki/Terminal_mode).
+    fn unset_raw_mode(&mut self, fd: RawFd);
+    /// Spawn a new terminal, with an optional file to open in a terminal program.
     fn spawn_terminal(&mut self, file_to_open: Option<PathBuf>) -> (RawFd, RawFd);
-    fn read_from_tty_stdout(&mut self, pid: RawFd, buf: &mut [u8]) -> Result<usize, nix::Error>;
-    fn write_to_tty_stdin(&mut self, pid: RawFd, buf: &mut [u8]) -> Result<usize, nix::Error>;
-    fn tcdrain(&mut self, pid: RawFd) -> Result<(), nix::Error>;
-    fn kill(&mut self, pid: RawFd) -> Result<(), nix::Error>;
+    /// Read bytes from the standard output of the virtual terminal referred to by `fd`.
+    fn read_from_tty_stdout(&mut self, fd: RawFd, buf: &mut [u8]) -> Result<usize, nix::Error>;
+    /// Write bytes to the standard input of the virtual terminal referred to by `fd`.
+    fn write_to_tty_stdin(&mut self, fd: RawFd, buf: &mut [u8]) -> Result<usize, nix::Error>;
+    /// Wait until all output written to the object referred to by `fd` has been transmitted.
+    fn tcdrain(&mut self, fd: RawFd) -> Result<(), nix::Error>;
+    /// Terminate the
+    // FIXME `RawFd` is semantically the wrong type here. It should either be a raw libc::pid_t,
+    // or a nix::unistd::Pid.
```

Comment:
I'm not sure I understand why?

Reply:
see `man kill.2`, `man kill.3` and the signature of `nix::sys::signal::kill`. All three take an argument called `pid` of type `pid_t` (just a `typedef`ed C `int` in reality), rather than a file descriptor. [`RawFd`](https://doc.rust-lang.org/std/os/unix/io/type.RawFd.html) happens to also be a `typedef`ed C `int`, so it works technically, but semantically it makes no sense: signals aren't sent to entities associated to file descriptors, they are sent to processes, which are associated to `pid`s.

Reply:
Good catch - that means there's an error in the `spawn_terminal` method as well. It took quite some time to get that one to work properly and I went through quite some iterations of the return types while doing so. I assume the erroneous return value was from one of them.